### PR TITLE
[no-release-info] Add additional tests for manipulating large JSON documents and fix corner case bugs in JSON_LOOKUP and JSON_INSERT

### PR DESCRIPTION
**Pull request by nicktobey**
This PR adds additional tests for calling JSON_INSERT on large JSON documents. It also fixes three issues with IndexedJsonDocuments:

1) Some operations are not supported by the new optimized implementation for JSON_LOOKUP, such as wildcards on array paths (eg `$[*]`). Instead of returning an error, we detect the error and fall back on the original implementation.

2) Attempting to insert a value into a document could cause an infinite loop.

3) We would fail to read some keys from an IndexedJsonDocument's StaticMap if the document contained arrays.

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@nicktobey __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>7879aa0</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>7879aa0</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d54e94c</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d54e94c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@nicktobey __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>7f66c4b</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>7f66c4b</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/store/prolly/tree/json_indexed_document.go [hunk 1]
```diff
@@ -107,29 +106,40 @@ func getBytesFromIndexedJsonMap(ctx context.Context, m StaticJsonMap) (bytes []b
...
-			return nil, err
-		}
-		nonIndexedDoc := types.JSONDocument{Val: val}
-		return nonIndexedDoc.Lookup(ctx, pathString)
-	}
-	result, err := i.tryLookup(ctx, pathString)
-	if err == unknownLocationKeyError {
-		if sqlCtx, ok := ctx.(*sql.Context); ok {
-			sqlCtx.GetLogger().Warn(err)
+func tryWithFallback(
```
Comment (fulghum) [inline go/store/prolly/tree/json_indexed_document.go:109]: Nice clean pattern with this function! 👏

File: go/store/prolly/tree/json_indexed_document.go [hunk 2]
```diff
@@ -107,29 +106,40 @@ func getBytesFromIndexedJsonMap(ctx context.Context, m StaticJsonMap) (bytes []b
...
+	ctx context.Context,
+	i IndexedJsonDocument,
+	tryFunc func() error,
+	fallbackFunc func(document types.JSONDocument) error) error {
+	err := tryFunc()
+	if err == unknownLocationKeyError || err == unsupportedPathError {
+		if err == unknownLocationKeyError {
+			if sqlCtx, ok := ctx.(*sql.Context); ok {
+				sqlCtx.GetLogger().Warn(err)
```
Comment (fulghum) [inline go/store/prolly/tree/json_indexed_document.go:118]: Do we already have tests that cover using unsupported path features and test that the fallback works correctly and that we log a session warning?

**Review by fulghum — APPROVED** [review]
Looks good!
